test/gtest/common/test_helpers.cc

```diff
@@ -15,6 +15,134 @@ namespace ucs {
 
 const double test_timeout_in_sec = 60.;
 
+const double watchdog_timeout_default = 900.; // 15 minutes
+
+static test_watchdog_t watchdog;
+
+void *watchdog_func(void *arg)
+{
+    int ret = 0;
+    double now;
+    struct timespec timeout;
+
+    pthread_mutex_lock(&watchdog.mutex);
+
+    /* sync with parent thread */
+    pthread_barrier_wait(&watchdog.barrier);
```

Comment:
isn't it dangerous to wait for barrier while mutex is held (potential deadlock)?

Reply:
yes, this is dangerous, but the `mutex+barrier` pair (instead of `barrier+mutex`) prevents the following situation:
0. let's assume that gtest is executed on a single core machine (but the situation below can occur even in multi core machine under high load)
1. main thread calls `watchdog_start()` ... watchdog thread is up and running:
watchdog thread calls `pthread_barrier_wait` and then main thread call `pthread_barrier_wait` -- main thread continues execution
2. main thread runs some test (e.g. empty test that will call `watchdog_signal()` in ~0 seconds).
3. then watchdog thread acquires `mutex` and go to `pthread_cond_timedwait()` - so it may occur that watchdog won't receive the signal and will sleep until time out.

In the current implementation:
1. watchdog acquires `mutex` and then calls `pthread_barrier_wait()` and main thread calls `pthread_barrier_wait()` w/o `mutex` held.
2. after `barrier()`, main thread can call `watchdog_signal()` right after the `barrier()` and this is guaranteed that watchdog will receive the signal. btw, `watchdog_start()` calls `watchdog_signal()` to test this situation

is it OK?